File: src/storage.rs
use async_trait::async_trait;
use aws_config::meta::region::RegionProviderChain;
use aws_sdk_s3::{
    error::{GetObjectError, GetObjectErrorKind},
    presigning::config::PresigningConfig,
    ByteStream, Client, SdkError,
};

use super::retreiver;
use std::{fs, io::Write, time::Duration};

// Strategy パターンっぽくしたかったが思ったようにはなっていない。
// （ Rust にデザインパターンをそのまま当てようとするのが間違っている感）
#[async_trait]
pub trait Storage {
    async fn load(&self) -> Vec<retreiver::SeismicIntensityStation>;
    async fn save(&self, stations: &Vec<retreiver::SeismicIntensityStation>);
}

pub struct FileStorage;
#[async_trait]
impl Storage for FileStorage {
    async fn load(&self) -> Vec<super::retreiver::SeismicIntensityStation> {
        // FIXME: 初期実行（ファイルが存在しない）を想定する
        let stations_str = fs::read_to_string("stations.json").unwrap();
        let stations = serde_json::from_str(stations_str.as_str()).unwrap();
        stations
    }

    async fn save(&self, stations: &Vec<super::retreiver::SeismicIntensityStation>) {
        let stations_str = serde_json::to_string(&stations).unwrap();

        let mut file = fs::File::create("stations.json").unwrap();
        file.write_all(stations_str.as_bytes()).unwrap();
    }
}

pub struct S3Storage {
    bucket: String,
    key: String,
    csv_key: String,
    client: Client,
}

impl S3Storage {
    pub async fn new(bucket: String) -> Self {
        let region_provider = RegionProviderChain::default_provider().or_else("ap-northeast-1");
        let config = aws_config::from_env().region(region_provider).load().await;
        let client = Client::new(&config);

        Self {
            bucket: bucket,
            key: "stations.json".to_string(),
            csv_key: "Stations.csv".to_string(),
            client: client,
        }
    }

    pub async fn generate_csv_presigned_url(&self) -> String {
        let presigning_config =
            PresigningConfig::expires_in(Duration::from_secs(24 * 60 * 60 * 7)).unwrap();
        let resp = self
            .client
            .get_object()
            .bucket(self.bucket.to_string())
            .key(self.csv_key.to_string())
            .presigned(presigning_config)
            .await
            .unwrap();

        resp.uri().to_string()
    }

    pub async fn save_csv(&self, csv: String) {
        let bytestream = ByteStream::from(csv.as_bytes().to_vec());
        self.client
            .put_object()
            .bucket(self.bucket.to_string())
            .key(self.csv_key.to_string())
            .body(bytestream)
            .send()
            .await
            .unwrap();
    }
}

#[async_trait]
impl Storage for S3Storage {
    async fn load(&self) -> Vec<retreiver::SeismicIntensityStation> {
        let resp = self
            .client
            .get_object()
            .bucket(self.bucket.to_string())
            .key(self.key.to_string())
            .send()
            .await;

        match resp {
            Ok(_) => {
                let bytes = resp.unwrap().body.collect().await.unwrap().into_bytes();
                let text = String::from_utf8_lossy(bytes.as_ref());

                return serde_json::from_str(text.as_ref()).unwrap();
            }
            Err(SdkError::ServiceError {
                err:
                    GetObjectError {
                        kind: GetObjectErrorKind::NoSuchKey(_),
                        ..
                    },
                ..
            }) => {
                return vec![];
            }
            _ => {}
        };

        panic!();
    }

    async fn save(&self, stations: &Vec<retreiver::SeismicIntensityStation>) {
        let stations_str = serde_json::to_string(&stations).unwrap();
        let bytestream = ByteStream::from(stations_str.as_bytes().to_vec());

        self.client
            .put_object()
            .bucket(self.bucket.to_string())
            .key(self.key.to_string())
            .body(bytestream)
            .send()
            .await
            .unwrap();
    }
}

File: internal/storage/storage.go
package storage

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"io"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	s3types "github.com/aws/aws-sdk-go-v2/service/s3/types"

	"github.com/p2pquake/stations_watcher/internal/retreiver"
)

const (
	defaultRegion = "ap-northeast-1"
	jsonKey       = "stations.json"
	csvKey        = "Stations.csv"
)

type S3Storage struct {
	bucket  string
	jsonKey string
	csvKey  string
	client  *s3.Client
}

func NewS3Storage(ctx context.Context, bucket string) (*S3Storage, error) {
	cfg, err := loadConfig(ctx)
	if err != nil {
		return nil, err
	}
	return &S3Storage{
		bucket:  bucket,
		jsonKey: jsonKey,
		csvKey:  csvKey,
		client:  s3.NewFromConfig(cfg),
	}, nil
}

func loadConfig(ctx context.Context) (aws.Config, error) {
	cfg, err := config.LoadDefaultConfig(ctx, config.WithRegion(defaultRegion))
	if err != nil {
		return aws.Config{}, err
	}
	if cfg.Region == "" {
		cfg.Region = defaultRegion
	}
	return cfg, nil
}

func (s *S3Storage) Load(ctx context.Context) ([]retreiver.SeismicIntensityStation, error) {
	resp, err := s.client.GetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(s.jsonKey),
	})
	if err != nil {
		var nsk *s3types.NoSuchKey
		if errors.As(err, &nsk) {
			return []retreiver.SeismicIntensityStation{}, nil
		}
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var stations []retreiver.SeismicIntensityStation
	if err := json.Unmarshal(body, &stations); err != nil {
		return nil, err
	}
	return stations, nil
}

func (s *S3Storage) Save(ctx context.Context, stations []retreiver.SeismicIntensityStation) error {
	data, err := json.Marshal(stations)
	if err != nil {
		return err
	}
	_, err = s.client.PutObject(ctx, &s3.PutObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(s.jsonKey),
		Body:   bytes.NewReader(data),
	})
	return err
}

func (s *S3Storage) SaveCSV(ctx context.Context, csv string) error {
	_, err := s.client.PutObject(ctx, &s3.PutObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(s.csvKey),
		Body:   bytes.NewReader([]byte(csv)),
	})
	return err
}

func GetObjectBytes(ctx context.Context, bucket, key string) ([]byte, error) {
	cfg, err := loadConfig(ctx)
	if err != nil {
		return nil, err
	}
	client := s3.NewFromConfig(cfg)
	resp, err := client.GetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
	})
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	return io.ReadAll(resp.Body)
}
